cunumeric/deferred.py

```diff
@@ -1108,10 +1109,18 @@ def _diag_helper(
                 # get slice of the original array by the offset
                 if offset > 0:
                     matrix = matrix.slice(start + 1, slice(offset, None))
-                if matrix.shape[n - 1] < matrix.shape[n]:
-                    diag = diag.promote(start + 1, matrix.shape[ndim - 1])
+                if trace:
+                    if matrix.ndim == 2:
+                        diag = diag.promote(0, matrix.shape[0])
+                        diag = diag.project(1, 0).promote(1, matrix.shape[1])
```

Comment:
Do you need to remove dimension 1 here before promoting? I think `diag` starts out with `ndim == 0`, and after the above statement it now has `ndim == 1` (i.e. it doesn't have a dimension 1).

Reply:
for this case I do following in array.py:
```
         if trace:
                if N != 2:
                    raise ValueError(
                        " exactly 2 axes should be passed to trace"
                    )
                if self.ndim == 2:
                    out_shape = (1,)
                elif self.ndim > 2:
                    out_shape = tr_shape
                else:
                    raise ValueError(
                        "dimension of the array for trace operation:"
                        " should be >=2"
                    )
```

Reply:
You're right, my mistake